### PR TITLE
Ropes issue5

### PR DESCRIPTION
**Pull request by dweinflash**
Allow images to be filtered according to tag name. Clicking on the tag link of an image on the Index or Show page displays a filtered list of images, all with at least one matching tag.

### CODE REVIEW

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -1,5 +1,37 @@
...
+  def test_filter_same_tags
+    valid_image_url = 'https://www.gstatic.com/webp/gallery3/1.png'
+    tag = 'test'
+    Image.create!(url: valid_image_url, tag_list: tag)
+    get "/images/?tag=#{tag}"
+    assert_select '.js-image-card', 1
+    Image.create!(url: valid_image_url, tag_list: tag)
+    get "/images/?tag=#{tag}"
+    assert_select '.js-image-card', 2
+  end
```
Comment (wanyue0714) [inline test/controllers/images_controller_test.rb:33]: Could we also add a test for the tag param doesn't match with any images' tags? sth like this:
```
tag_many = %w[tag1 tag2]
Image.create!(url: valid_image_url, tag_list: %w[tag1, tag2])
get images_url(tag: 'tag3')
assert_select '.js-image-card', count: 0
```

File: test/controllers/images_controller_test.rb [hunk 2]
```diff
@@ -1,5 +1,37 @@
 require 'test_helper'
 class ImagesControllerTest < ActionDispatch::IntegrationTest
+  def test_filter_many_tags
+    valid_image_url = 'https://www.gstatic.com/webp/gallery3/1.png'
```
Comment (wanyue0714) [inline test/controllers/images_controller_test.rb:4]: If we use this valid_image_url is used all these tags tests, then we could define it as a constant at the top, so it could be re-used
```
require 'test_helper'

VALID_IMAGE_URL = 'https://www.gstatic.com/webp/gallery3/1.png'.freeze

def test_filter_many_tags
    tag_1 = 'tag1'
    tag_many = %w[tag1 tag2]
    Image.create!(url: VALID_IMAGE_URL, tag_list: tag_many)
```

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -1,5 +1,37 @@
...
+    tag = 'test'
+    Image.create!(url: valid_image_url, tag_list: tag)
+    get "/images/?tag=#{tag}"
+    assert_select '.js-image-card', 1
+    Image.create!(url: valid_image_url, tag_list: tag)
+    get "/images/?tag=#{tag}"
+    assert_select '.js-image-card', 2
+  end
+
```
Comment (wanyue0714) [inline test/controllers/images_controller_test.rb:34]: Also, might be good to test if the tags is empty, it does render all images, which is to cover 
```
if @images.empty?
   @images = Image.all.order(created_at: :desc)
else
```

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -1,6 +1,11 @@
 class ImagesController < ActionController::Base
   def index
-    @images = Image.all.order(created_at: :desc)
+    image_tags = params.permit(:tag)[:tag]
+    if image_tags.nil?
```
[the remark this answers is not in the document]
Comment (dweinflash) [reply app/controllers/images_controller.rb:4]: I tried making this change and it unfortunately breaks a few of our tests. Surprisingly, image_tags can be nil, as is the case in `test_index` from `images_controller_test` and `test_home` from `application_controller_test`.

[the remark this answers is not in the document]
Comment (dweinflash) [reply app/controllers/images_controller.rb:4]: I think image_tags is nil when we use `get images_path` and `get root_path`?

File: test/controllers/images_controller_test.rb (hunk 3 above)
Comment (dweinflash) [reply test/controllers/images_controller_test.rb:34]: Sounds good! Adding a test to show that all images are rendered when filtering by an empty tag

File: app/views/images/show.html.erb [hunk 5]
```diff
@@ -11,7 +11,9 @@
       <% if @image %>
         <p>
           Tag: <br>
-          <span class="js_image_tag"><%= @image&.tag_list.join(', ') %></span>
+          <span class="js_image_tag">
+            <%= raw(@image&.tag_list.map { |t| link_to t, images_path(tag: t) }.join(', ')) %>
```
[the remark this answers is not in the document]
Comment (dweinflash) [reply app/views/images/show.html.erb:15]: I think we need `raw` in order to avoid escaping the string? Without `raw` my link displays on the page as _<a href="/images?tag=dog">dog</a>_

File: app/controllers/images_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (wanyue0714) [reply app/controllers/images_controller.rb:4]: Aha, okay, correct my comment, we'd better to cover empty string like '', when the url is `/images/?tag=''` it should also go into the first case(rendering all images) right?
So we could use `unless image_tags.present?` to include all cases like `nil, ''`

File: app/views/images/show.html.erb (hunk 5 above)
[the remark this answers is not in the document]
Comment (wanyue0714) [reply app/views/images/show.html.erb]: Hahaha, I just realized I wrote the same thing in my ropes 1 year ago. Totally forgot that 😆
Okay, so using raw works fine, you're right!
Rails escapes html string because it might be malicious code, imagine that pop up some malicious window. But if you tell Rails that it's `html_safe`, it'll pass it through. So normally we just call `.html_safe`, what `raw` doing is converting the string to a string, and then calling html_safe. So if it's already a string, we could call html_safe directly, it skips an unnecessary step and makes code easy to understand
